torch/csrc/distributed/spmd/engine.cpp

```diff
@@ -0,0 +1,136 @@
+#include <torch/csrc/distributed/spmd/engine.h>
+
+#include <torch/csrc/distributed/spmd/event_impl.h>
+
+namespace torch {
+namespace distributed {
+namespace spmd {
+
+/////////////////////////////////////////////////////////////////////
+//                           RootHandler                           //
+/////////////////////////////////////////////////////////////////////
+
+std::vector<EventSchema> RootHandler::ingressEvents() const {
+  return {};
+}
+
+std::vector<EventSchema> RootHandler::egressEvents() const {
+  return {
+      EventType::PREPARE_MODULE,
+      EventType::PRE_FORWARD,
+      EventType::POST_FORWARD};
+}
+
+std::vector<std::shared_ptr<Future>> RootHandler::handleEvent(
+    const c10::intrusive_ptr<Event>& /* unused */) {
+  TORCH_INTERNAL_ASSERT(
+      false, "RootHandler should not handle any ingress Events.");
+}
+
+/////////////////////////////////////////////////////////////////////
+//                             Engine                              //
+/////////////////////////////////////////////////////////////////////
+
+Engine::Engine(std::vector<std::shared_ptr<EventHandler>> handlers)
+    : handlers_(std::move(handlers)) {
+  // temporary helper data structure
+  std::unordered_map<
+      EventSchema,
+      std::vector<std::shared_ptr<HandlerNode>>,
+      EventSchema::Hash>
+      ingressMap;
+
+  std::vector<std::shared_ptr<HandlerNode>> handlerNodes;
+  // RootHandler generates Type I events
+  auto rootHandler =
+      std::make_shared<HandlerNode>(std::make_shared<RootHandler>());
+  handlerNodes.push_back(rootHandler);
+
+  {
+    // BUILD GRAPH
+
+    // check ingress events, and build ingressMap that points from each event
+    // to it's corresponding handlers.
+    for (auto& handler : handlers_) {
+      auto handlerNode = std::make_shared<HandlerNode>(handler);
+      handlerNodes.push_back(handlerNode);
+      for (auto& eventSchema : handler->ingressEvents()) {
+        // TODO: check no duplicated events
+        ingressMap[eventSchema].push_back(handlerNode);
+      }
+    }
+
+    // check egress events to build the bipartite graph
+    for (auto& handlerNode : handlerNodes) {
+      for (auto& eventSchema : handlerNode->handler_->egressEvents()) {
+        auto iter = eventNodes_.find(eventSchema);
+        std::shared_ptr<EventNode> eventNode;
+        if (iter == eventNodes_.end()) {
+          eventNode = std::make_shared<EventNode>(eventSchema);
+          eventNodes_.emplace(eventSchema, eventNode);
+        } else {
+          eventNode = iter->second;
+        }
+
+        handlerNode->nextEdges_.push_back(eventNode);
+        for (auto& nextHandlerNode : ingressMap[eventSchema]) {
+          eventNode->nextEdges_.push_back(nextHandlerNode);
+        }
+      }
+    }
+  }
+
+  {
+    // VERIFY GRAPH
```

Comment:
Does the second `Bucketer` mean delayed allreduce? Is this the only case that you think we need a cycle here? If so, I don't think we need to introduce cycles in the graph.

We can remove the 2nd Bucketer event, i.e.:
Bucketer -> BUCKET_READY -> AllReduceComm -> COMM_DONE -> GLOBAL_GRAD_READY

As long as we for each event, we keep an attribute of `iter_` or `epoch_`, so an event can use different strategies across different iterations. DDP comm hook already has this support for hybrid gradient compression in temporal dimension. We can have similar support for Bucketer.

If we want to be even more flexible, then each event can be viewed as a list of per-time-window events, each of which corresponds an event of the same type for a non-overlapping time window.

Reply:
> Does the second Bucketer mean delayed allreduce? 

That's the event handler. 

With the current event handling graph, we cannot directly go from event to event. An event will need to pass some event handling operator to get transformed into another event. This is similar to Apache Beam's PCollection (Event) and Transform (Handler). Almost all graph-based processing systems adopted this design, e.g., Twitter Storm, Spark Streaming, Linkedin Samza, Flink, etc. 

The reason that it needs to come back to the same bucketer is that Bucketer remembers the mapping from grads to bucket offsets. To get rid of this, and attach states to events and splits the current bucketer to two, or we can have a dedicated concept for internal states. This will avoid the cycle in this PR. 

I am thinking does this mean we should re-work our handler API a bit and let each handler consumes exactly one event?